

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Welcome to ProjectM1S1Bioinfo’s documentation! &mdash; ProjectM1S1Bioinfo 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home" alt="Documentation Home"> ProjectM1S1Bioinfo
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Welcome to ProjectM1S1Bioinfo’s documentation!</a><ul>
<li><a class="reference internal" href="#documentation-for-the-code">Documentation for the Code</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-blast_hitter">BlastHitter</a></li>
<li><a class="reference internal" href="#module-clusterizer">Clusterizer</a></li>
<li><a class="reference internal" href="#module-refseq_scraper">RefSeqScraper</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</div>
            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">ProjectM1S1Bioinfo</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
        
      <li>Welcome to ProjectM1S1Bioinfo’s documentation!</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-projectm1s1bioinfo-s-documentation">
<h1>Welcome to ProjectM1S1Bioinfo’s documentation!<a class="headerlink" href="#welcome-to-projectm1s1bioinfo-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<div class="section" id="documentation-for-the-code">
<h2>Documentation for the Code<a class="headerlink" href="#documentation-for-the-code" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
</div>
</div>
</div>
<div class="section" id="module-blast_hitter">
<span id="blasthitter"></span><h1>BlastHitter<a class="headerlink" href="#module-blast_hitter" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="blast_hitter.BlastHitter">
<em class="property">class </em><code class="sig-prename descclassname">blast_hitter.</code><code class="sig-name descname">BlastHitter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">query</span></em>, <em class="sig-param"><span class="n">subject</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a class for BLASTing two genomes against each other and
for deducing reciprocal best hits that are common between the two 
blastp results files.</p>
<p>Attributes</p>
<dl class="simple">
<dt>query_path<span class="classifier">str</span></dt><dd><p>the relative path for the query genome file on the disk.</p>
</dd>
<dt>query_name<span class="classifier">str</span></dt><dd><p>the name of the query genome.</p>
</dd>
<dt>subject_path<span class="classifier">str</span></dt><dd><p>the relative path for the subject genome file on the disk.</p>
</dd>
<dt>query_name<span class="classifier">str</span></dt><dd><p>the name of the subject genome.</p>
</dd>
<dt>results_dir<span class="classifier">str</span></dt><dd><p>the relative path of the results repository for saving output files.</p>
</dd>
<dt>genomes_dir<span class="classifier">str</span></dt><dd><p>the relative path of the genomes repository.</p>
</dd>
<dt>first_blastp<span class="classifier">str</span></dt><dd><p>the path of the first .blastp file (query vs. subject blast).</p>
</dd>
<dt>second_blastp<span class="classifier">str</span></dt><dd><p>the path of the second .blastp file (subject vs. query blast).</p>
</dd>
<dt>rbh<span class="classifier">str</span></dt><dd><p>the path of the reciprocal best hits file (.blastp tabulated format).</p>
</dd>
</dl>
<dl class="py method">
<dt id="blast_hitter.BlastHitter.best_hits_from_blast">
<em class="property">static </em><code class="sig-name descname">best_hits_from_blast</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blastp_file</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.best_hits_from_blast" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the best hit for every protein based on the blast output
file. The algorithm is simple, take the first appearance of the 
protein as the best hit because BLAST sends back the best hits in
descending order of quality.</p>
<p>Parameters</p>
<dl class="simple">
<dt>blastp_file<span class="classifier">str</span></dt><dd><p>The blast output file with a tabulated format (outfmt = 6).</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>besthits_dict<span class="classifier">dict</span></dt><dd><p>a dictionary that contains all protein queries of a given genome,
with our query as a key, and the corresponding best hit as a value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.bidir_best_hits">
<em class="property">static </em><code class="sig-name descname">bidir_best_hits</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blastp1</span></em>, <em class="sig-param"><span class="n">blastp2</span></em>, <em class="sig-param"><span class="n">out</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.bidir_best_hits" title="Permalink to this definition">¶</a></dt>
<dd><p>This function determines the reciprocal best hits between two blast
files, it calls the class’s best_hits_from_blast static method to 
determine the best hits for every blast file, creates two sets of 
tuples; query and best hit for the first blast, best hit and query
for the second blast (inverses the second blast file). it then 
proceeds with the intersection between our two sets in order to
establish the reciprocal best hits (RBH).</p>
<p>Parameters</p>
<dl class="simple">
<dt>blastp1<span class="classifier">str</span></dt><dd><p>The First blast file path.</p>
</dd>
<dt>blastp2<span class="classifier">str</span></dt><dd><p>The second blast file path.</p>
</dd>
<dt>out<span class="classifier">str</span></dt><dd><p>The reciprocal best hits output file path, tabulated (outfmt = 6).</p>
</dd>
</dl>
<p>Returns</p>
<p>None.
Creates an out RBH file based on the first blast file after 
calculating the reciprocal best hits.</p>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.blast_them">
<code class="sig-name descname">blast_them</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.blast_them" title="Permalink to this definition">¶</a></dt>
<dd><p>This method launches two BLASTs, our two genomes against each other.
it uses the class’s static methods defind earlier.</p>
<p>Returns</p>
<dl class="simple">
<dt>str</dt><dd><p>The first blast output file path.</p>
</dd>
<dt>str</dt><dd><p>The second blast output file path.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.evalue_dist">
<em class="property">static </em><code class="sig-name descname">evalue_dist</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blastp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.evalue_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates an evalue distribution plot from a given blastp file.</p>
<p>Parameters</p>
<dl class="simple">
<dt>blastp<span class="classifier">str</span></dt><dd><p>The blast file path to be analyzed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.from_list">
<em class="property">classmethod </em><code class="sig-name descname">from_list</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prots_list</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>This class method instantiates BlastHitter objects from a list of
proteomes/genomes.</p>
<p>Parameters</p>
<dl class="simple">
<dt>prots_list<span class="classifier">list</span></dt><dd><p>a list of genomes/proteoms file paths.</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>list</dt><dd><p>a list of BlastHitter objects.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.getRbh">
<code class="sig-name descname">getRbh</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.getRbh" title="Permalink to this definition">¶</a></dt>
<dd><p>The getter of the RBH file path for a certain blasthitter object.</p>
<p>Returns</p>
<dl class="simple">
<dt>str</dt><dd><p>RBH file path.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.parse_fasta">
<em class="property">static </em><code class="sig-name descname">parse_fasta</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proteome</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.parse_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>This function parses a genome/proteome a file.</p>
<p>Parameters</p>
<dl class="simple">
<dt>proteome<span class="classifier">str</span></dt><dd><p>the path of the genome/proteome file.</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>seqdic<span class="classifier">dict             </span></dt><dd><p>dictionary with protein accessions as keys and the corresponding
fasta sequences as values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.rbh_them">
<code class="sig-name descname">rbh_them</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.rbh_them" title="Permalink to this definition">¶</a></dt>
<dd><p>This method calculates the reciprocal best hits between our two 
genomes after BLASTing them, it creates the RBH file.</p>
<p>Returns</p>
<dl class="simple">
<dt>str</dt><dd><p>The RBH output file path.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.seqkit_stat">
<em class="property">static </em><code class="sig-name descname">seqkit_stat</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proteome</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.seqkit_stat" title="Permalink to this definition">¶</a></dt>
<dd><p>This function prints important information about the proteome
that will be analyzed. It parses the genome first then prints the 
number of sequences, cumulated length, minimum, average and maximum 
length. It mimics seqkit stats output.</p>
<p>Parameters</p>
<dl class="simple">
<dt>proteome<span class="classifier">str</span></dt><dd><p>the path of the proteome file.</p>
</dd>
</dl>
<p>Returns</p>
<p>None.</p>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.universal_blast">
<em class="property">static </em><code class="sig-name descname">universal_blast</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">query</span></em>, <em class="sig-param"><span class="n">subject</span></em>, <em class="sig-param"><span class="n">out</span></em>, <em class="sig-param"><span class="n">outfmt</span><span class="o">=</span><span class="default_value">6</span></em>, <em class="sig-param"><span class="n">typ</span><span class="o">=</span><span class="default_value">'p'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.universal_blast" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the blast command to be executed in the terminal.</p>
<p>Parameters</p>
<dl class="simple">
<dt>query<span class="classifier">str</span></dt><dd><p>query proteome file path.</p>
</dd>
<dt>subject<span class="classifier">str</span></dt><dd><p>subject proteome file path.</p>
</dd>
<dt>out<span class="classifier">str</span></dt><dd><p>blast output file path.</p>
</dd>
<dt>outfmt<span class="classifier">int, optional</span></dt><dd><p>blast results format. The default tabulated without headers = 6.</p>
</dd>
<dt>typ<span class="classifier">str, optional</span></dt><dd><p>type of blast to run. The default is blastp.</p>
</dd>
</dl>
<p>Returns</p>
<p>str</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-clusterizer">
<span id="clusterizer"></span><h1>Clusterizer<a class="headerlink" href="#module-clusterizer" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="clusterizer.Clusterizer">
<em class="property">class </em><code class="sig-prename descclassname">clusterizer.</code><code class="sig-name descname">Clusterizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blasthitters</span></em>, <em class="sig-param"><span class="n">proteomes</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a class for clusterizing groups of orthologue proteines (OG),
extracting their respective sequences from the corresponding given 
proteome, aligning each cluster individually and then constructing the 
super-alignement.</p>
<p>Attributes</p>
<dl class="simple">
<dt>rbh_files<span class="classifier">list</span></dt><dd><p>A list of reciprocal best hits file paths generated after each blast
hitter object.</p>
</dd>
<dt>proteomes<span class="classifier">list</span></dt><dd><p>A list of all proteome paths present in a given analysis.</p>
</dd>
<dt>working_cluster<span class="classifier">dictionary</span></dt><dd><p>A collection of clusters (protein accession numbers) present among 
the given RBH files, filtering was applied to only get a maximum of
one protein per species inside a cluster (no paralogues). 
Dictionary values correspond to cluster IDs which are auto-incremented
int values.</p>
</dd>
<dt>corr_species_cluster<span class="classifier">dictionary </span></dt><dd><p>The corresponding species cluster for the given working cluster 
dictionary, it’s the same data structure as the working cluster except
the fact that protein accession numbers are replaced with Species 
of origin name.</p>
</dd>
<dt>super_alignement<span class="classifier">str</span></dt><dd><p>the file path of the calculated super-alignement in aligned fasta 
format (.afa).</p>
</dd>
</dl>
<dl class="py method">
<dt id="clusterizer.Clusterizer.all_pairs_rbh">
<em class="property">static </em><code class="sig-name descname">all_pairs_rbh</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">files</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.all_pairs_rbh" title="Permalink to this definition">¶</a></dt>
<dd><p>This function take multiple RBH files and returns a long list of
all RBH couples that are present in those FILES, it calls the previous
class static method.</p>
<p>Parameters</p>
<dl class="simple">
<dt>files<span class="classifier">list</span></dt><dd><p>a list of all RBH file paths to be analyzed.</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>total<span class="classifier">list</span></dt><dd><p>a list of tuples of all RBH couples in all specified files.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.cluster_from_proteome">
<em class="property">static </em><code class="sig-name descname">cluster_from_proteome</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster</span></em>, <em class="sig-param"><span class="n">proteomes</span></em>, <em class="sig-param"><span class="n">out</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.cluster_from_proteome" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a fasta file for a given cluster, it firsts parses all 
proteome files to generate one big dictionary, it then filters it
to take only the headers and the sequence that correspond to accession
numbers found in a cluster.</p>
<p>Parameters</p>
<dl class="simple">
<dt>cluster<span class="classifier">tuple</span></dt><dd><p>A tuple of protein accession numbers found in a cluster.</p>
</dd>
<dt>proteomes<span class="classifier">list</span></dt><dd><p>A list of all proteomes file path</p>
</dd>
<dt>out<span class="classifier">str</span></dt><dd><p>The output file path.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.cluster_them">
<code class="sig-name descname">cluster_them</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.cluster_them" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is used for generalizing the clustering workflow,
First, clusters from rbh files and write them out to a text. Second,
generate the species cluste, third, filter the dictionaries to only
one protein per orgranism inside a cluster and lastly, write out the
filtered clusters to a file and set the two dictionaries as class 
properties so they can be used elsewise.</p>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.clustering">
<em class="property">static </em><code class="sig-name descname">clustering</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rbh_files</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.clustering" title="Permalink to this definition">¶</a></dt>
<dd><p>The main clustering algorithm to find reciprocal best hits among
multiple RBH blast files. it uses the class’s all_pairs_rbh static
method as a starting point. After getting the list of all RBH couples
in all RBH files (a list of tuples), it traverses it and throws the 
couple of RBHs in a set inside a list if one of them is present in a
given set, otherwise if the loop exited without a break, create 
another cluster from the two RBHs, Time complexity O(n²).</p>
<p>Parameters</p>
<dl class="simple">
<dt>rbh_files<span class="classifier">list</span></dt><dd><p>A list of RBH blast file paths.</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>dict</dt><dd><p>a dictionary of all clusters present among all RBH files,
keys are auto incremented integers which correspond to cluster IDs,
values are NCBI accession numbers of the proteins present in a
given cluster.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.clusters_to_txt">
<em class="property">static </em><code class="sig-name descname">clusters_to_txt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_dict</span></em>, <em class="sig-param"><span class="n">out</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.clusters_to_txt" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a cluster dictionary to a text file, each line corresponds
to a cluster of orthologue proteins.</p>
<p>Parameters</p>
<dl class="simple">
<dt>cluster_dict<span class="classifier">dict</span></dt><dd><p>a dictionary of cluster ids and cluster accessions.</p>
</dd>
<dt>out<span class="classifier">str</span></dt><dd><p>the output text file path.</p>
</dd>
</dl>
<p>Returns</p>
<p>None.</p>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.draw_tree">
<code class="sig-name descname">draw_tree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">iters</span><span class="o">=</span><span class="default_value">10</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.draw_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>This method generates an interactive phylogenetic tree from a
super alignement file. It calculates a newick tree with RAxML and then
visualizes it using ETEtoolkit.</p>
<p>Parameters</p>
<dl class="simple">
<dt>iters<span class="classifier">int, optional</span></dt><dd><p>Number of repetitions for bootstrap. The default is 10.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.max_one_species_per_cluster">
<em class="property">static </em><code class="sig-name descname">max_one_species_per_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_species</span></em>, <em class="sig-param"><span class="n">cluster_dict</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.max_one_species_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters cluster dictionaries to only one protein per species.
It scans the cluster species dictionary and sends back the subset of it
for which we only have one protein per species, then in filters the
cluster accession dictionary with the ids of the filtered one.</p>
<p>Parameters</p>
<dl class="simple">
<dt>cluster_species<span class="classifier">dict</span></dt><dd><p>The cluster dictionary of species names.</p>
</dd>
<dt>cluster_dict<span class="classifier">dict</span></dt><dd><p>The cluster dictionary (with accessions).</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>dict</dt><dd><p>The filtered cluster accession dicionary.</p>
</dd>
<dt>filtered<span class="classifier">dict</span></dt><dd><p>the filtered cluster species dictionary.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.muscle">
<em class="property">static </em><code class="sig-name descname">muscle</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_dict</span></em>, <em class="sig-param"><span class="n">proteomes</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.muscle" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates fasta files for each cluster, and then aligns each of them
using MUSCLE.</p>
<p>Parameters</p>
<dl class="simple">
<dt>cluster_dict<span class="classifier">dict</span></dt><dd><p>The cluster dictionary (with accessions).</p>
</dd>
<dt>proteomes<span class="classifier">list</span></dt><dd><p>The list of all proteomes file path for a cluster.</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>afasta_files<span class="classifier">list</span></dt><dd><p>A list of all the generated aligned fasta file paths (.afa).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.one_align_to_rule_them_all">
<code class="sig-name descname">one_align_to_rule_them_all</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.one_align_to_rule_them_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Generating the super alignement file for the analyzed species of 
interest, it firsts generates MSAs for all clusters using MUSCLE,
and then it concatenates them. all results will be saved to respective
directories. It alse sent the super alignement file path as a class
property.</p>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.pair_rbh">
<em class="property">static </em><code class="sig-name descname">pair_rbh</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.pair_rbh" title="Permalink to this definition">¶</a></dt>
<dd><p>This function parses the reciprocal best hits file and returns a 
list of all rbh couples. it extracts the first and the second columns.</p>
<p>Parameters</p>
<dl class="simple">
<dt>file<span class="classifier">str</span></dt><dd><p>The RBH file_path.</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>list</dt><dd><p>a list of tuples of all RBH couples.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.species_cluster">
<em class="property">static </em><code class="sig-name descname">species_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_dict</span></em>, <em class="sig-param"><span class="n">proteomes</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.species_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>This function generates a species cluster from an accession
cluster, keys are cluster ids (auto incremented int values) and values
are clusters of species names for every protein in the cluster.
First it parses all proteome files and every header in each proteome,
afterwords, it scans the already generated cluster dictionary to
generate a replica of it swapping the accession number by the species
of origin.</p>
<p>Parameters</p>
<dl class="simple">
<dt>cluster_dict<span class="classifier">dict</span></dt><dd><p>The cluster dictionary (with accessions).</p>
</dd>
<dt>proteomes<span class="classifier">list</span></dt><dd><p>A list of proteomes file paths.</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>dict</dt><dd><p>The corresponding species dictionary to a given cluster accession
dictionary.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.super_alignement">
<em class="property">static </em><code class="sig-name descname">super_alignement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_dict</span></em>, <em class="sig-param"><span class="n">cluster_species</span></em>, <em class="sig-param"><span class="n">maligns</span></em>, <em class="sig-param"><span class="n">out</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.super_alignement" title="Permalink to this definition">¶</a></dt>
<dd><p>This function concatenates all the multiple alignement files into
one super-alignement fasta file, each header is the name of the species
studied and the sequence corresponds to the concatenation of all the
clusters. It first creates a seed dictionary of the first cluster i.e;
with keys as species’ name and values as the aligned sequences. 
Afterwards, and for each cluster, it appends the aligned sequence into
the species name if the cluster has a protein from a given species, 
otherwise it fills it with gaps that has the same length as the multi-
ple alignement inside a cluster.</p>
<p>Parameters</p>
<dl class="simple">
<dt>cluster_species<span class="classifier">dict</span></dt><dd><p>The cluster dictionary of species names.</p>
</dd>
<dt>cluster_dict<span class="classifier">dict</span></dt><dd><p>The cluster dictionary (with accessions).</p>
</dd>
<dt>maligns<span class="classifier">list</span></dt><dd><p>A list of all multiple alignement files paths.</p>
</dd>
<dt>out<span class="classifier">str</span></dt><dd><p>The super-alignement output file path.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.tree_generator">
<em class="property">static </em><code class="sig-name descname">tree_generator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">super_alignement</span></em>, <em class="sig-param"><span class="n">bootstrap</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.tree_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a newick tree from a given super-alignement. it uses
RAxML’s maximum likelihood algorithm.</p>
<p>Parameters</p>
<dl class="simple">
<dt>super_alignement<span class="classifier">str</span></dt><dd><p>The super-alignement file path.</p>
</dd>
<dt>bootstrap<span class="classifier">int</span></dt><dd><p>Number of repetitions for bootstrap.</p>
</dd>
</dl>
<p>Returns</p>
<dl class="simple">
<dt>tree_name<span class="classifier">str</span></dt><dd><p>The newick tree to be visualized (the file path).</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-refseq_scraper">
<span id="refseqscraper"></span><h1>RefSeqScraper<a class="headerlink" href="#module-refseq_scraper" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Ghassan DABANE, Megane BOUJEANT

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>