

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Welcome to ProjectM1S1Bioinfo’s documentation! &mdash; ProjectM1S1Bioinfo 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home" alt="Documentation Home"> ProjectM1S1Bioinfo
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Welcome to ProjectM1S1Bioinfo’s documentation!</a><ul>
<li><a class="reference internal" href="#quickstart-demo-guiless">Quickstart Demo (GUIless)</a></li>
<li><a class="reference internal" href="#documentation">Documentation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-interface">Graphical User Interface</a></li>
<li><a class="reference internal" href="#module-refseq_scraper">RefSeqScraper</a></li>
<li><a class="reference internal" href="#module-blast_hitter">BlastHitter</a></li>
<li><a class="reference internal" href="#module-clusterizer">Clusterizer</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</div>
            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">ProjectM1S1Bioinfo</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
        
      <li>Welcome to ProjectM1S1Bioinfo’s documentation!</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-projectm1s1bioinfo-s-documentation">
<h1>Welcome to ProjectM1S1Bioinfo’s documentation!<a class="headerlink" href="#welcome-to-projectm1s1bioinfo-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<div class="section" id="quickstart-demo-guiless">
<h2>Quickstart Demo (GUIless)<a class="headerlink" href="#quickstart-demo-guiless" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p>Let’s import our two classes</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">blast_hitter</span> <span class="kn">import</span> <span class="n">BlastHitter</span>
<span class="kn">from</span> <span class="nn">clusterizer</span> <span class="kn">import</span> <span class="n">Clusterizer</span>
</pre></div>
</div>
</li>
<li><p>Proteomes can be downloaded with RefSeqScraper script, or manually saved to data/genomes repository. A list of all paths should be initalised, so that we can use BlastHitter’s factory method to create a bunch of blasthitter objects of all possible permutations.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">proteomes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;../data/genomes/Rickettsia_rickettsii_str._Arizona_strain=Arizona_protein.faa&quot;</span><span class="p">,</span>
<span class="s2">&quot;../data/genomes/Streptococcus_pneumoniae_R6_strain=R6_protein.faa&quot;</span><span class="p">,</span>
<span class="s2">&quot;../data/genomes/Streptococcus_pyogenes_strain=NCTC8232_protein.faa&quot;</span><span class="p">,</span>
<span class="s2">&quot;../data/genomes/Streptococcus_thermophilus_LMD-9_strain=LMD-9_protein.faa&quot;</span><span class="p">,</span>
<span class="s2">&quot;../data/genomes/Piscirickettsia_salmonis_strain=Psal-158_protein.faa&quot;</span><span class="p">]</span>

<span class="n">bhitters</span> <span class="o">=</span> <span class="n">BlastHitter</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="n">proteomes</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>We can blast them and accumulate the reciprocal best hits with a for loop :</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">bh</span> <span class="ow">in</span> <span class="n">bhitters</span>  <span class="p">:</span>
    <span class="n">bh</span><span class="o">.</span><span class="n">blast_them</span><span class="p">()</span>
    <span class="n">bh</span><span class="o">.</span><span class="n">rbh_them</span><span class="p">()</span>
</pre></div>
</div>
</li>
<li><p>Now let’s create a Clusterizer object after populating our blasthitters with RBH files :</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">clust</span> <span class="o">=</span> <span class="n">Clusterizer</span><span class="p">(</span><span class="n">bhitters</span><span class="p">,</span> <span class="n">proteomes</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p>The next and final step would be to create clusters, aligning each one and concatenating all of them and last but not least would be to launch the phylogenetic algorithm and to draw the newick tree:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">clust</span><span class="o">.</span><span class="n">cluster_them</span><span class="p">()</span>
<span class="n">clust</span><span class="o">.</span><span class="n">one_align_to_rule_them_all</span><span class="p">()</span>
<span class="n">clust</span><span class="o">.</span><span class="n">draw_tree</span><span class="p">()</span>
</pre></div>
</div>
</li>
</ul>
</div>
<div class="section" id="documentation">
<h2>Documentation<a class="headerlink" href="#documentation" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
</div>
</div>
</div>
<div class="section" id="module-interface">
<span id="graphical-user-interface"></span><h1>Graphical User Interface<a class="headerlink" href="#module-interface" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="interface.window">
<em class="property">class </em><code class="sig-prename descclassname">interface.</code><code class="sig-name descname">window</code><a class="headerlink" href="#interface.window" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a class for the  graphical user interface which offers different
functionalities.</p>
<p>It has 5 main functions :</p>
<blockquote>
<div><dl class="simple">
<dt>Make_a_Tree :</dt><dd><p>A feature that allows to make a phylogenetic tree from a couple of chosen
proteomes using BLAST/MUSCLE/RAxML.</p>
</dd>
<dt>Download :</dt><dd><p>A feature that permets the user to download proteomes from RefSeq.</p>
</dd>
<dt>E-values :</dt><dd><p>Visualizes the distribution of e-values from a BLASTp results file.</p>
</dd>
<dt>Statistics :</dt><dd><p>A feature that allows to see some statistics of a chosen proteome.</p>
</dd>
<dt>Exit :</dt><dd><p>Button in the menu to exit the graphical interface.</p>
</dd>
</dl>
</div></blockquote>
<dl class="py method">
<dt id="interface.window.Download_proteome">
<code class="sig-name descname">Download_proteome</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dropdown_list</span></em><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.Download_proteome" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is called when the user has chosen, therefore has selected, the proteome that he wants to download.
It downloads the proteome, chosen by the user, from RefSeq database to the disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dropdown_list</strong> (<em>list</em>) – A list that contains all names of proteomes available in RefSeq
database which contain the pattern written by the user.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="interface.window.Search_Download_proteome">
<code class="sig-name descname">Search_Download_proteome</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.Search_Download_proteome" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is called by the “Download” feature.
It asks  theuser to write a few letters in order to search this pattern in the list
of all proteomes available in RefSeq database.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><strong>pattern</strong> – pattern that contains a few letters write by user.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="interface.window.Select_Download_proteome">
<code class="sig-name descname">Select_Download_proteome</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pattern</span></em><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.Select_Download_proteome" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is called when user presses enter on his keyboard.
It display a dropdown list of all proteomes, that correspond to the pattern given by
the user, available in RefSeq database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pattern</strong> (<em>str</em>) – pattern that contains a few letters write by user.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>dropdown_list</strong> – A list that contains all names of proteomes available in RefSeq
database which contain the pattern written by the user.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="interface.window.createMenuBar">
<code class="sig-name descname">createMenuBar</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.createMenuBar" title="Permalink to this definition">¶</a></dt>
<dd><p>This method creates a menu bar that contains several tabs, each
tab gives access to a feature of the GUI.</p>
</dd></dl>

<dl class="py method">
<dt id="interface.window.distributions_of_evalues">
<code class="sig-name descname">distributions_of_evalues</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.distributions_of_evalues" title="Permalink to this definition">¶</a></dt>
<dd><p>This methid is called by the “E-values” feature.
It displays a list of all BLASTp results files that are available on the disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><strong>blast_presents</strong> – A list of all BLASTp files available on the disk.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="interface.window.proteomes_in_disk">
<code class="sig-name descname">proteomes_in_disk</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.proteomes_in_disk" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is called when the user wants the “Make_a_Tree” feature.
It displays list of all proteomes that are available in the disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><strong>p_presents</strong> – A list that contains all names of proteomes available in the disk.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="interface.window.reset">
<code class="sig-name descname">reset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.reset" title="Permalink to this definition">¶</a></dt>
<dd><p>This method allows widget reset, except the widget of the menu bar.
This function is particularly useful when user changes tabs,
or also if user clicks on the tab again.</p>
</dd></dl>

<dl class="py method">
<dt id="interface.window.stats">
<code class="sig-name descname">stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.stats" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is called by the “Statistics” feature.
It displays a list of all proteomes that are available on the disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><strong>prot_presents</strong> – A list of all proteomes available in the disk.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="interface.window.validate_blast_selection">
<code class="sig-name descname">validate_blast_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blast_presents</span></em><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.validate_blast_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is called when the user has selected the BLASTp file that he wants to visualize
as a distribution of e-values.
It displays a new window with a histogram of the distribution of e-values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>blast_presents</strong> (<em>list</em>) – A list of all BLASTp files available in the disk.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Opens a new window with a histogram of the distribution
of e-values of a chosen BLASTp file.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="interface.window.validate_prot_select">
<code class="sig-name descname">validate_prot_select</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prot_presents</span></em><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.validate_prot_select" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is called when the user has selected the proteome that
he wants to see some statistics about.
It displays a new window with the statistics of proteome that user has
selected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>prot_presents</strong> (<em>list</em>) – A list of all proteomes available in the disk.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Opens a new window with some statistics about the proteome
that the user has chosen.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="interface.window.validate_proteome_selection">
<code class="sig-name descname">validate_proteome_selection</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">p_presents</span></em><span class="sig-paren">)</span><a class="headerlink" href="#interface.window.validate_proteome_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is called when the user has to validate the proteome selection and he wants to obtain a phylogenetic tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>p_presents</strong> (<em>list</em>) – list that contains all names of proteomes available in the disk.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Opens a new window with the phylogenetic tree.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-refseq_scraper">
<span id="refseqscraper"></span><h1>RefSeqScraper<a class="headerlink" href="#module-refseq_scraper" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="refseq_scraper.RefSeqScraper">
<em class="property">class </em><code class="sig-prename descclassname">refseq_scraper.</code><code class="sig-name descname">RefSeqScraper</code><a class="headerlink" href="#refseq_scraper.RefSeqScraper" title="Permalink to this definition">¶</a></dt>
<dd><p>This class will scan refseq summary assembly dataframe, pick only latest
and complete genomes and then will create a new column ‘readable’ that is
the combination of organism and infraspecific name columns.
based on refseq’s assembly summary file :
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt">ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt</a>
A new column “readable” will be created with pandas to make it easier for the user to
choose a certain genome.</p>
<dl class="py attribute">
<dt id="refseq_scraper.RefSeqScraper.data">
<code class="sig-name descname">data</code><a class="headerlink" href="#refseq_scraper.RefSeqScraper.data" title="Permalink to this definition">¶</a></dt>
<dd><p>A pandas dataframe that contains all refseq genomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.df</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="refseq_scraper.RefSeqScraper.cart">
<code class="sig-name descname">cart</code><a class="headerlink" href="#refseq_scraper.RefSeqScraper.cart" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of species names to download their genomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="refseq_scraper.RefSeqScraper.add_to_cart">
<code class="sig-name descname">add_to_cart</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">species</span></em><span class="sig-paren">)</span><a class="headerlink" href="#refseq_scraper.RefSeqScraper.add_to_cart" title="Permalink to this definition">¶</a></dt>
<dd><p>This function adds a species to the object’s cart.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>species</strong> (<em>str</em>) – the name of the species to be downloaded, should match
records in the dataframe.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A sentence that confirms the addition of a given species to the cart.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="refseq_scraper.RefSeqScraper.download_genome">
<code class="sig-name descname">download_genome</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#refseq_scraper.RefSeqScraper.download_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads all genomes from the ftps paths in the cart and saves
them to data/genomes directory.</p>
</dd></dl>

<dl class="py method">
<dt id="refseq_scraper.RefSeqScraper.mine_ftps">
<code class="sig-name descname">mine_ftps</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#refseq_scraper.RefSeqScraper.mine_ftps" title="Permalink to this definition">¶</a></dt>
<dd><p>This function searches the dataframe and extracts the ftps that
correspond to species in the cart.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of ftp paths.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="refseq_scraper.RefSeqScraper.mine_species">
<code class="sig-name descname">mine_species</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#refseq_scraper.RefSeqScraper.mine_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Asks the user to give a few letters and finds a few records in the
dataframe that correspond to the input pattern, if the user would like
to add something that he sees, it does it.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-blast_hitter">
<span id="blasthitter"></span><h1>BlastHitter<a class="headerlink" href="#module-blast_hitter" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="blast_hitter.BlastHitter">
<em class="property">class </em><code class="sig-prename descclassname">blast_hitter.</code><code class="sig-name descname">BlastHitter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">query</span></em>, <em class="sig-param"><span class="n">subject</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a class for BLASTing two genomes against each other and
for deducing reciprocal best hits that are common between the two
blastp results files.</p>
<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.query_path">
<code class="sig-name descname">query_path</code><a class="headerlink" href="#blast_hitter.BlastHitter.query_path" title="Permalink to this definition">¶</a></dt>
<dd><p>the relative path for the query genome file on the disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.query_name">
<code class="sig-name descname">query_name</code><a class="headerlink" href="#blast_hitter.BlastHitter.query_name" title="Permalink to this definition">¶</a></dt>
<dd><p>the name of the query genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.subject_path">
<code class="sig-name descname">subject_path</code><a class="headerlink" href="#blast_hitter.BlastHitter.subject_path" title="Permalink to this definition">¶</a></dt>
<dd><p>the relative path for the subject genome file on the disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.subject_name">
<code class="sig-name descname">subject_name</code><a class="headerlink" href="#blast_hitter.BlastHitter.subject_name" title="Permalink to this definition">¶</a></dt>
<dd><p>the name of the subject genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.results_dir">
<code class="sig-name descname">results_dir</code><a class="headerlink" href="#blast_hitter.BlastHitter.results_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>the relative path of the results repository for saving output files.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.genomes_dir">
<code class="sig-name descname">genomes_dir</code><a class="headerlink" href="#blast_hitter.BlastHitter.genomes_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>the relative path of the genomes repository.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.first_blastp">
<code class="sig-name descname">first_blastp</code><a class="headerlink" href="#blast_hitter.BlastHitter.first_blastp" title="Permalink to this definition">¶</a></dt>
<dd><p>the path of the first .blastp file (query vs. subject blast).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.second_blastp">
<code class="sig-name descname">second_blastp</code><a class="headerlink" href="#blast_hitter.BlastHitter.second_blastp" title="Permalink to this definition">¶</a></dt>
<dd><p>the path of the second .blastp file (subject vs. query blast).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="blast_hitter.BlastHitter.rbh">
<code class="sig-name descname">rbh</code><a class="headerlink" href="#blast_hitter.BlastHitter.rbh" title="Permalink to this definition">¶</a></dt>
<dd><p>the path of the reciprocal best hits file (.blastp tabulated format).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.best_hits_from_blast">
<em class="property">static </em><code class="sig-name descname">best_hits_from_blast</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blastp_file</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.best_hits_from_blast" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the best hit for every protein based on the blast output
file. The algorithm is simple, take the first appearance of the
protein as the best hit because BLAST sends back the best hits in
descending order of quality.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>blastp_file</strong> (<em>str</em>) – The blast output file with a tabulated format (outfmt = 6).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>besthits_dict</strong> – A dictionary that contains all protein queries of a given genome,
with our query as a key, and the corresponding best hit as a value.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.bidir_best_hits">
<em class="property">static </em><code class="sig-name descname">bidir_best_hits</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blastp1</span></em>, <em class="sig-param"><span class="n">blastp2</span></em>, <em class="sig-param"><span class="n">out</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.bidir_best_hits" title="Permalink to this definition">¶</a></dt>
<dd><p>This function determines the reciprocal best hits between two blast
files, it calls the class’s best_hits_from_blast static method to
determine the best hits for every blast file, creates two sets of
tuples; query and best hit for the first blast, best hit and query
for the second blast (inverses the second blast file). it then
proceeds with the intersection between our two sets in order to
establish the reciprocal best hits (RBH).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>blastp1</strong> (<em>str</em>) – The First blast file path.</p></li>
<li><p><strong>blastp2</strong> (<em>str</em>) – The second blast file path.</p></li>
<li><p><strong>out</strong> (<em>str</em>) – The reciprocal best hits output file path, tabulated (outfmt = 6).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Creates an out RBH file based on the first blast file after
calculating the reciprocal best hits.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.blast_them">
<code class="sig-name descname">blast_them</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.blast_them" title="Permalink to this definition">¶</a></dt>
<dd><p>This method launches two BLASTs, our two genomes against each other.
it uses the class’s static methods defind earlier.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><ul class="simple">
<li><p><em>str</em> – The first blast output file path.</p></li>
<li><p><em>str</em> – The second blast output file path.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.evalue_dist">
<em class="property">static </em><code class="sig-name descname">evalue_dist</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blastp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.evalue_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates an evalue distribution plot from a given blastp file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>blastp</strong> (<em>str</em>) – The blast file path to be analyzed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.from_list">
<em class="property">classmethod </em><code class="sig-name descname">from_list</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">prots_list</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>This class method instantiates BlastHitter objects from a list of
proteomes/genomes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>prots_list</strong> (<em>list</em>) – a list of genomes/proteoms file paths.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of BlastHitter objects.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.getRbh">
<code class="sig-name descname">getRbh</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.getRbh" title="Permalink to this definition">¶</a></dt>
<dd><p>The getter of the RBH file path for a certain blasthitter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RBH file path.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.parse_fasta">
<em class="property">static </em><code class="sig-name descname">parse_fasta</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proteome</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.parse_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>This function parses a genome/proteome a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>proteome</strong> (<em>str</em>) – the path of the genome/proteome file.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>seqdic</strong> – dictionary with protein accessions as keys and the corresponding
fasta sequences as values.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.rbh_them">
<code class="sig-name descname">rbh_them</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.rbh_them" title="Permalink to this definition">¶</a></dt>
<dd><p>This method calculates the reciprocal best hits between our two
genomes after BLASTing them, it creates the RBH file.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>The RBH output file path.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.seqkit_stat">
<em class="property">static </em><code class="sig-name descname">seqkit_stat</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">proteome</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.seqkit_stat" title="Permalink to this definition">¶</a></dt>
<dd><p>This function prints important information about the proteome
that will be analyzed. It parses the genome first then prints the
number of sequences, cumulated length, minimum, average and maximum
length. It mimics seqkit stats output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>proteome</strong> (<em>str</em>) – the path of the proteome file.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Prints some statistics from a chosen proteome file</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="blast_hitter.BlastHitter.universal_blast">
<em class="property">static </em><code class="sig-name descname">universal_blast</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">query</span></em>, <em class="sig-param"><span class="n">subject</span></em>, <em class="sig-param"><span class="n">out</span></em>, <em class="sig-param"><span class="n">outfmt</span><span class="o">=</span><span class="default_value">6</span></em>, <em class="sig-param"><span class="n">typ</span><span class="o">=</span><span class="default_value">'p'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#blast_hitter.BlastHitter.universal_blast" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the blast command to be executed in the terminal.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> (<em>str</em>) – query proteome file path.</p></li>
<li><p><strong>subject</strong> (<em>str</em>) – subject proteome file path.</p></li>
<li><p><strong>out</strong> (<em>str</em>) – blast output file path.</p></li>
<li><p><strong>outfmt</strong> (<em>int</em><em>, </em><em>optional</em>) – blast results format. The default tabulated without headers = 6.</p></li>
<li><p><strong>typ</strong> (<em>str</em><em>, </em><em>optional</em>) – type of blast to run. The default is blastp.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The blast command to be executed.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-clusterizer">
<span id="clusterizer"></span><h1>Clusterizer<a class="headerlink" href="#module-clusterizer" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="clusterizer.Clusterizer">
<em class="property">class </em><code class="sig-prename descclassname">clusterizer.</code><code class="sig-name descname">Clusterizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">blasthitters</span></em>, <em class="sig-param"><span class="n">proteomes</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a class for clusterizing groups of orthologue proteines (OG),
extracting their respective sequences from the corresponding given
proteome, aligning each cluster individually and then constructing the
super-alignement.</p>
<dl class="py attribute">
<dt id="clusterizer.Clusterizer.rbh_files">
<code class="sig-name descname">rbh_files</code><a class="headerlink" href="#clusterizer.Clusterizer.rbh_files" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of reciprocal best hits file paths generated after each blast
hitter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="clusterizer.Clusterizer.proteomes">
<code class="sig-name descname">proteomes</code><a class="headerlink" href="#clusterizer.Clusterizer.proteomes" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of all proteome paths present in a given analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="clusterizer.Clusterizer.working_cluster">
<code class="sig-name descname">working_cluster</code><a class="headerlink" href="#clusterizer.Clusterizer.working_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>A collection of clusters (protein accession numbers) present among
the given RBH files, filtering was applied to only get a maximum of
one protein per species inside a cluster (no paralogues).
Dictionary values correspond to cluster IDs which are auto-incremented
int values.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="clusterizer.Clusterizer.corr_species_cluster">
<code class="sig-name descname">corr_species_cluster</code><a class="headerlink" href="#clusterizer.Clusterizer.corr_species_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>The corresponding species cluster for the given working cluster
dictionary, it’s the same data structure as the working cluster except
the fact that protein accession numbers are replaced with Species
of origin name.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="clusterizer.Clusterizer.super_alignement">
<code class="sig-name descname">super_alignement</code><a class="headerlink" href="#clusterizer.Clusterizer.super_alignement" title="Permalink to this definition">¶</a></dt>
<dd><p>the file path of the calculated super-alignement in aligned fasta
format (.afa).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.all_pairs_rbh">
<em class="property">static </em><code class="sig-name descname">all_pairs_rbh</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">files</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.all_pairs_rbh" title="Permalink to this definition">¶</a></dt>
<dd><p>This function take multiple RBH files and returns a long list of
all RBH couples that are present in those FILES, it calls the previous
class static method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>files</strong> (<em>list</em>) – a list of all RBH file paths to be analyzed.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>total</strong> – a list of tuples of all RBH couples in all specified files.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.cluster_from_proteome">
<em class="property">static </em><code class="sig-name descname">cluster_from_proteome</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster</span></em>, <em class="sig-param"><span class="n">proteomes</span></em>, <em class="sig-param"><span class="n">out</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.cluster_from_proteome" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a fasta file for a given cluster, it firsts parses all
proteome files to generate one big dictionary, it then filters it
to take only the headers and the sequence that correspond to accession
numbers found in a cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster</strong> (<em>tuple</em>) – A tuple of protein accession numbers found in a cluster.</p></li>
<li><p><strong>proteomes</strong> (<em>list</em>) – A list of all proteomes file path</p></li>
<li><p><strong>out</strong> (<em>str</em>) – The output file path.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.cluster_them">
<code class="sig-name descname">cluster_them</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.cluster_them" title="Permalink to this definition">¶</a></dt>
<dd><p>This method is used for generalizing the clustering workflow,
First, clusters from rbh files and write them out to a text. Second,
generate the species cluste, third, filter the dictionaries to only
one protein per orgranism inside a cluster and lastly, write out the
filtered clusters to a file and set the two dictionaries as class
properties so they can be used elsewise.</p>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.clustering">
<em class="property">static </em><code class="sig-name descname">clustering</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rbh_files</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.clustering" title="Permalink to this definition">¶</a></dt>
<dd><p>The main clustering algorithm to find reciprocal best hits among
multiple RBH blast files. it uses the class’s all_pairs_rbh static
method as a starting point. After getting the list of all RBH couples
in all RBH files (a list of tuples), it traverses it and throws the
couple of RBHs in a set inside a list if one of them is present in a
given set, otherwise if the loop exited without a break, create
another cluster from the two RBHs, Time complexity O(n²).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>rbh_files</strong> (<em>list</em>) – A list of RBH blast file paths.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary of all clusters present among all RBH files,
keys are auto incremented integers which correspond to cluster IDs,
values are NCBI accession numbers of the proteins present in a
given cluster.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.clusters_to_txt">
<em class="property">static </em><code class="sig-name descname">clusters_to_txt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_dict</span></em>, <em class="sig-param"><span class="n">out</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.clusters_to_txt" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes a cluster dictionary to a text file, each line corresponds
to a cluster of orthologue proteins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_dict</strong> (<em>dict</em>) – A dictionary of cluster ids and cluster accessions.</p></li>
<li><p><strong>out</strong> (<em>str</em>) – The output text file path.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.draw_tree">
<code class="sig-name descname">draw_tree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">iters</span><span class="o">=</span><span class="default_value">10</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.draw_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>This method generates an interactive phylogenetic tree from a
super alignement file. It calculates a newick tree with RAxML and then
visualizes it using ETEtoolkit.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>iters</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of repetitions for bootstrap. The default is 10.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.max_one_species_per_cluster">
<em class="property">static </em><code class="sig-name descname">max_one_species_per_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_species</span></em>, <em class="sig-param"><span class="n">cluster_dict</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.max_one_species_per_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters cluster dictionaries to only one protein per species.
It scans the cluster species dictionary and sends back the subset of it
for which we only have one protein per species, then in filters the
cluster accession dictionary with the ids of the filtered one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_species</strong> (<em>dict</em>) – The cluster dictionary of species names.</p></li>
<li><p><strong>cluster_dict</strong> (<em>dict</em>) – The cluster dictionary (with accessions).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>dict</em> – The filtered cluster accession dicionary.</p></li>
<li><p><strong>filtered</strong> (<em>dict</em>) – the filtered cluster species dictionary.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.muscle">
<em class="property">static </em><code class="sig-name descname">muscle</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_dict</span></em>, <em class="sig-param"><span class="n">proteomes</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.muscle" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates fasta files for each cluster, and then aligns each of them
using MUSCLE.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_dict</strong> (<em>dict</em>) – The cluster dictionary (with accessions).</p></li>
<li><p><strong>proteomes</strong> (<em>list</em>) – The list of all proteomes file path for a cluster.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>afasta_files</strong> – A list of all the generated aligned fasta file paths (.afa).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.one_align_to_rule_them_all">
<code class="sig-name descname">one_align_to_rule_them_all</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.one_align_to_rule_them_all" title="Permalink to this definition">¶</a></dt>
<dd><p>Generating the super alignement file for the analyzed species of
interest, it firsts generates MSAs for all clusters using MUSCLE,
and then it concatenates them. all results will be saved to respective
directories. It alse sent the super alignement file path as a class
property.</p>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.pair_rbh">
<em class="property">static </em><code class="sig-name descname">pair_rbh</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">file</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.pair_rbh" title="Permalink to this definition">¶</a></dt>
<dd><p>This function parses the reciprocal best hits file and returns a
list of all rbh couples. it extracts the first and the second columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file</strong> (<em>str</em>) – The RBH file_path.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of tuples of all RBH couples.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.species_cluster">
<em class="property">static </em><code class="sig-name descname">species_cluster</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_dict</span></em>, <em class="sig-param"><span class="n">proteomes</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.species_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>This function generates a species cluster from an accession
cluster, keys are cluster ids (auto incremented int values) and values
are clusters of species names for every protein in the cluster.
First it parses all proteome files and every header in each proteome,
afterwords, it scans the already generated cluster dictionary to
generate a replica of it swapping the accession number by the species
of origin.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_dict</strong> (<em>dict</em>) – The cluster dictionary (with accessions).</p></li>
<li><p><strong>proteomes</strong> (<em>list</em>) – A list of proteomes file paths.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The corresponding species dictionary to a given cluster accession
dictionary.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="id0">
<em class="property">static </em><code class="sig-name descname">super_alignement</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cluster_dict</span></em>, <em class="sig-param"><span class="n">cluster_species</span></em>, <em class="sig-param"><span class="n">maligns</span></em>, <em class="sig-param"><span class="n">out</span></em><span class="sig-paren">)</span><a class="headerlink" href="#id0" title="Permalink to this definition">¶</a></dt>
<dd><p>This function concatenates all the multiple alignement files into
one super-alignement fasta file, each header is the name of the species
studied and the sequence corresponds to the concatenation of all the
clusters. It first creates a seed dictionary of the first cluster i.e;
with keys as species’ name and values as the aligned sequences.
Afterwards, and for each cluster, it appends the aligned sequence into
the species name if the cluster has a protein from a given species,
otherwise it fills it with gaps that has the same length as the multi-
ple alignement inside a cluster.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_species</strong> (<em>dict</em>) – The cluster dictionary of species names.</p></li>
<li><p><strong>cluster_dict</strong> (<em>dict</em>) – The dictionary of cluster accession numbers.</p></li>
<li><p><strong>maligns</strong> (<em>list</em>) – A list of all multiple alignement files paths.</p></li>
<li><p><strong>out</strong> (<em>str</em>) – The super-alignement output file path.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="clusterizer.Clusterizer.tree_generator">
<em class="property">static </em><code class="sig-name descname">tree_generator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">super_alignement</span></em>, <em class="sig-param"><span class="n">bootstrap</span></em><span class="sig-paren">)</span><a class="headerlink" href="#clusterizer.Clusterizer.tree_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a newick tree from a given super-alignement. it uses
RAxML’s maximum likelihood algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>super_alignement</strong> (<em>str</em>) – The super-alignement file path.</p></li>
<li><p><strong>bootstrap</strong> (<em>int</em>) – Number of repetitions for bootstrap.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>tree_name</strong> – The newick tree to be visualized (the file path).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Ghassan DABANE, Megane BOUJEANT

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>